lib/LaTeXML/Package/TeX.pool.ltxml

```diff
@@ -6612,48 +6626,76 @@ DefConstructor('\lx@add@Preamble@PI Undigested',
 
 #======================================================================
 
-DefConstructor('\DUAL OptionalKeyVals {}{}',
+DefKeyVal('lx@DUAL', 'reversion',              'UndigestedKey');
+DefKeyVal('lx@DUAL', 'content_reversion',      'UndigestedKey');
+DefKeyVal('lx@DUAL', 'presentation_reversion', 'UndigestedKey');
+our @PENDING_DUAL_XMARGS = ();
+DefConstructor('\DUAL OptionalKeyVals:lx@DUAL {}{}',
   "<ltx:XMDual role='&GetKeyVal(#1,role)'>#2<ltx:XMWrap>#3</ltx:XMWrap></ltx:XMDual>",
-  reversion => sub {
-    my ($whatsit) = @_;
-    my ($kv, $content, $presentation) = $whatsit->getArgs;
-    if (!$LaTeXML::DUAL_BRANCH) {
-      Tokens(T_CS('\DUAL'),
-        $whatsit->getDefinition->getParameters->revertArguments($kv, $content, $presentation)); }
-    elsif (($LaTeXML::DUAL_BRANCH eq 'content')
-      || !GetKeyVal($kv, 'hide_content_reversion')) {
-      Revert($content); }
-    elsif ($LaTeXML::DUAL_BRANCH eq 'presentation') {
-      Revert($presentation); }
-    else {
-      Warn('misdefined', 'DUAL', $_[0],
-        "$LaTeXML::DUAL_BRANCH Unknown DUAL_BRANCH: $LaTeXML::DUAL_BRANCH");
-      Tokens(T_CS('\DUAL'),
-        $whatsit->getDefinition->getParameters->revertArguments($kv, $content, $presentation)); }
-  },
-  sizer => '#3');    # size according to presentation
-
-sub addMeaningRec {
-  my ($document, $node, $meaning) = @_;
-  if ($node->nodeType == XML_ELEMENT_NODE) {
-    my $qname = $document->getModel->getNodeQName($node);
-    if ($qname eq 'ltx:XMArg') { }    # DONT cross through into arguments!
-    elsif ($qname eq 'ltx:XMTok') {
-      if ((($node->getAttribute('role') || 'UNKNOWN') eq 'UNKNOWN')
-        && !$node->getAttribute('meaning')) {
-        $document->setAttribute($node, meaning => $meaning); } }
-    else {
-      foreach my $c ($node->childNodes) {
-        addMeaningRec($document, $c, $meaning); } } }
-  return; }
+  beforeDigest => sub {
+    push(@PENDING_DUAL_XMARGS, {});
+    return; },
+  afterDigest => sub {
+    my ($stomach, $whatsit) = @_;
+    my $kv     = $whatsit->getArg(1);
+    my $xmargs = pop(@PENDING_DUAL_XMARGS);
+    $whatsit->setProperties($kv->getPairs);
+    if (!$kv->getValue('reversion')) {
+      if (!$kv->getValue('content_reversion')) {
+        $whatsit->setProperty('content_reversion' => Tokens(T_PARAM, T_OTHER('2'))); }
+      if (!$kv->getValue('presentation_reversion')) {
+        $whatsit->setProperty('presentation_reversion' => Tokens(T_PARAM,
+            ($kv->getValue('hide_content_reversion') ? T_OTHER('3') : T_OTHER('2')))); } }
+    $whatsit->setProperties(%$xmargs);    # Hopefully no name class with XM<digits>
+    return; },
+  sizer => '#3');                         # size according to presentation
+
+# Construction aids
+# Build an XMDual (via \DUAL) given the content & presentation forms.
+# These forms are provided as Tokens, invoking the appropriate constructor macros,
+# and refering to any arguments using #1, #2.... (see T_XMArg for syntactic sugar)
```

Comment:
You're using `T_XMArg` already? I was ambivalent about creating such a trivial function, but it helped the readability in physics.sty so much it was worth it.  The main thing is that it's short; the `T_` prefix was only because it almost was just a token (2 actually). And the `XMArg` part seemed pretty specific to the application.  Or not?

Reply:
Let's ignore my `T_OUT_PARAM` comment for now, given the current direction of the PR, maybe the overlap is less than expected. Can always revisit later.